### PR TITLE
EASY-1640 command line incorrect fetch.txt file results in unclear error messaging

### PR DESCRIPTION
**Pull request by stephanHogenboom**
Fixes EASY-1640

#### When applied it will...
* log a more user friendly error

#### Related pull requests on github

repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.bagstore/component/FileSystemComponent.scala [hunk 1]
```diff
@@ -224,6 +224,8 @@ trait FileSystemComponent extends DebugEnhancedLogging {
             fileId <- id.toFileId
             location <- toRealLocation(fileId)
           } yield (bagDir.toAbsolutePath.resolve(item.path), location)
+        } recoverWith {
+          case e: Exception => Failure(new IllegalArgumentException(s"Bag-id found in fetch.txt can not be found in the bag-store: ${ e.getMessage }"))
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.bagstore/component/FileSystemComponent.scala]: I'm afraid there are more causes for exceptions so make the pattern match more specific.

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 2]
```diff
@@ -167,4 +170,26 @@ class BagStoreSpec extends TestSupportFixture
...
+      .fromFile(fetchFilePath.toFile)
+      .getLines()
+      .map(line => line.replaceAll("01", "10"))
+      .foreach(printWriter.println)
+
+    printWriter.close()
+    tmpFile.renameTo(fetchFilePath.toFile)
+
+    bagStore.add(testBagPrunedB, Some(uuid1)) should matchPattern {
+      case Failure(ce: CompositeException) if ce.throwables.head.getMessage contains exceptionSuffixMessage =>
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: Try something like 
```suggestion
       case Failure(CompositeException(errors)) if erros.head.getMessage contains exceptionSuffixMessage =>
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 3]
```diff
@@ -167,4 +170,26 @@ class BagStoreSpec extends TestSupportFixture
...
+    val tmpFile = testBagPrunedB.resolve("fetch2.txt").toFile
+    val fetchFilePath = testBagPrunedB.resolve("fetch.txt")
+    val exceptionSuffixMessage = "Bag-id found in fetch.txt can not be found in the bag-store:"
+    val printWriter = new PrintWriter(tmpFile)
+
+    Source
+      .fromFile(fetchFilePath.toFile)
+      .getLines()
+      .map(line => line.replaceAll("01", "10"))
+      .foreach(printWriter.println)
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: If tmpFile is a better.File:
```suggestion
      .foreach(tmpFile.write)
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 4]
```diff
@@ -167,4 +170,26 @@ class BagStoreSpec extends TestSupportFixture
...
+
+    printWriter.close()
+    tmpFile.renameTo(fetchFilePath.toFile)
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: Or even shorter change the whole lot into something like
```scala
fetchFile.write(fetchFile.contentAsString.replaceAll("01", "10"))
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 5]
```diff
@@ -167,4 +169,16 @@ class BagStoreSpec extends TestSupportFixture
...
+
+  it should "result in a failure when the fetch.txt is invalid" in {
+    val uuid1 = UUID.fromString("00000000-0000-0000-0000-000000000001")
+    val fetchFile = File(testBagPrunedB.resolve("fetch.txt"))
+    val exceptionMessageSuffix = "Bag-id found in fetch.txt can not be found in the bag-store:"
+
+    Files.write(fetchFile.path, fetchFile.contentAsString.replaceAll("01", "10").getBytes())
+
+    bagStore.add(testBagPrunedB, Some(uuid1)) should matchPattern {
+      case Failure(CompositeException(errors)) if errors.head.getMessage contains exceptionMessageSuffix =>
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: to satisfy codacy:
```suggestion
       case Failure(CompositeException(errors)) if errors.headOption.getOrElse(new Exception()).getMessage contains exceptionMessageSuffix =>
```
Perhaps extract method for the condition.

File: src/main/scala/nl.knaw.dans.easy.bagstore/component/FileSystemComponent.scala [hunk 6]
```diff
@@ -224,6 +224,8 @@ trait FileSystemComponent extends DebugEnhancedLogging {
             fileId <- id.toFileId
             location <- toRealLocation(fileId)
           } yield (bagDir.toAbsolutePath.resolve(item.path), location)
+        } recoverWith {
+          case nsfe: NoSuchFileException => Failure(new IllegalArgumentException(s"Bag-id found in fetch.txt can not be found in the bag-store: ${ nsfe.getMessage }"))
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.bagstore/component/FileSystemComponent.scala]: The error message is not correct. This was not correctly formulated in the issue. It should be:
```suggestion
          case nsfe: NoSuchFileException => Failure(new IllegalArgumentException(s"Local-file-uri found in fetch.txt can not be found in the bag-store: ${ nsfe.getMessage }"))
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 7]
```diff
@@ -167,4 +169,20 @@ class BagStoreSpec extends TestSupportFixture
...
+
+    Files.write(fetchFile.path, fetchFile.contentAsString.replaceAll("01", "10").getBytes())
+
+    bagStore.add(testBagPrunedB, Some(uuid1)) should matchPattern {
+       case Failure(CompositeException(errors)) if oneOfTheErrorsContainsMessage(exceptionMessageSuffix, errors) =>
+    }
+  }
+
+  private def oneOfTheErrorsContainsMessage(exceptionMessageSuffix: String, errors: List[Throwable]) = {
+    errors.headOption.getOrElse(new Exception()).getMessage contains exceptionMessageSuffix
```
Comment (janvanmansum) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: Is there a cleaner way to do this?

**Review by janvanmansum — CHANGES_REQUESTED** [review]
* Text of error message is incorrect. This was incorrectly specified in the issue.
* Have another look at the unit test to see if you can make it a bit cleaner.

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala (hunk 7 above)
Comment (stephanHogenboom) [reply src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: This indeed looks a bit sloppy, i refined the code.

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 8]
```diff
@@ -167,4 +169,15 @@ class BagStoreSpec extends TestSupportFixture
...
   }
+
+  it should "result in a failure when the fetch.txt is invalid" in {
+    val uuid1 = UUID.fromString("00000000-0000-0000-0000-000000000001")
+    val fetchFile = File(testBagPrunedB.resolve("fetch.txt"))
+
+    Files.write(fetchFile.path, fetchFile.contentAsString.replaceAll("01", "10").getBytes())
+
+    bagStore.add(testBagPrunedB, Some(uuid1)) should matchPattern {
+       case Failure(CompositeException(errors)) if errors.exists(_.isInstanceOf[IllegalArgumentException]) =>
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: Could you make this test somewhat more precise, to see that it actually contains the excepted exception? Like, testing on the expected text.

**Review by rvanheest — CHANGES_REQUESTED** [review]

**Comment by stephanHogenboom** [comment]
@janvanmansum The text of the error message is updated and the test is made a bit cleaner

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 9]
```diff
@@ -167,4 +169,18 @@ class BagStoreSpec extends TestSupportFixture
...
     Files.getPosixFilePermissions(bagInStore.resolve("data/sub")) shouldBe dirPermissions
   }
+
+  it should "result in a failure when the fetch.txt is invalid" in {
+    val uuid1 = UUID.fromString("00000000-0000-0000-0000-000000000001")
+    val fetchFile = File(testBagPrunedB.resolve("fetch.txt"))
+
+    Files.write(fetchFile.path, fetchFile.contentAsString.replaceAll("01", "10").getBytes())
+
+    bagStore.add(testBagPrunedB, Some(uuid1)) should matchPattern {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: this is a bit of a weird test IMHO. Shouldn't it be something like:

```scala
class BagStoreSpec extends TestSupportFixture
  with OneInstancePerTest
  with Inspectors // new import! necessary for `forEvery` below
  with ...

// lots of tests

inside(bagStore.add(testBagPrunedB, Some(uuid1))) {
  case Failure(CompositeException(errors)) =>
    forEvery(errors)(e => {
      e shouldBe an[IllegalArgumentException]
      e.getMessage should include ("Local-file-uri found in fetch.txt can not be found in the bag-store: ")
    })
}
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 10]
```diff
@@ -17,12 +17,15 @@ package nl.knaw.dans.easy.bagstore.component
 
 import java.net.URI
 import java.nio.file.attribute.{ PosixFilePermission, PosixFilePermissions }
-import java.nio.file.{ Files, Path, Paths }
+import java.nio.file.{ Files, NoSuchFileException, Path, Paths }
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: unused import(s)?

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala (hunk 9 above)
Comment (stephanHogenboom) [reply src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala]: Nice, i was not aware of the forEvery function

File: src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala [hunk 11]
```diff
@@ -167,4 +169,19 @@ class BagStoreSpec extends TestSupportFixture
...
+
+  it should "result in a failure when the fetch.txt is invalid" in {
+    val uuid1 = UUID.fromString("00000000-0000-0000-0000-000000000001")
+    val fetchFile = File(testBagPrunedB.resolve("fetch.txt"))
+
+    Files.write(fetchFile.path, fetchFile.contentAsString.replaceAll("01", "10").getBytes())
+
+    inside(bagStore.add(testBagPrunedB, Some(uuid1))) {
+      case Failure(CompositeException(errors)) =>
+        forEvery(errors)(e => {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/component/BagStoreSpec.scala:182]:
```suggestion
        errors should have size 4
        forEvery(errors)(e => {
```

**Review by rvanheest — APPROVED** [review]
apart from the extra size check in the test, this looks good

**Pull request merged**